File: python-js/retention-assignments/algo-armstrong-numbers/js/armstrongNumbers.js
// How can you make this more scalable and reusable later?

function getDigits(number) {
    let ret = [0, 0, 0];
    if (number >= 100) {
        let val = Math.floor(number / 100);
        ret[2] = val;
        number -= val * 100;
    }
    if (number >= 10) {
        let val = Math.floor(number / 10);
        ret[1] = val;
        number -= val * 10;
    }
    if (number >= 1) {
        let val = Math.floor(number / 1);
        ret[0] = val;
    }
    return ret;
}

exports.findArmstrongNumbers = function(numbersList) {
    let ret = [];
    const singleDigitVals = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
    const doubleDigitVals = singleDigitVals.map(function(element) { return Math.pow(element, 2) });
    const tripleDigitVals = singleDigitVals.map(function(element) { return Math.pow(element, 3) });
    
    for (let i = 0; i < numbersList.length; i++) {
        let digitsList = getDigits(numbersList[i]);
        let sum = 0;
        // Calculate the sum based on armstrong number equation
        if (digitsList[2] > 0) {
            sum = tripleDigitVals[digitsList[2]] + tripleDigitVals[digitsList[1]] + tripleDigitVals[digitsList[0]];
        }
        else if (digitsList[1] > 0) {
            sum = doubleDigitVals[digitsList[2]] + doubleDigitVals[digitsList[1]] + doubleDigitVals[digitsList[0]];
        }
        else {
            sum = singleDigitVals[digitsList[2]] + singleDigitVals[digitsList[1]] + singleDigitVals[digitsList[0]];
        }
        // Check if the number is an armstrong number
        if (sum == numbersList[i]) {
            ret.push(sum);
        }
    }
    return ret;
};

File: python-js/intro-to-code/retention-assignments/algo-armstrong-numbers/python/armstrong_numbers.py
def get_digits(number):
    hundreds, tens, ones = 0, 0, 0
    if number >= 100:
        hundreds = int(number / 100)
        number -= hundreds * 100
    if number >= 10:
        tens = int(number / 10)
        number -= tens * 10
    if number >= 0:
        ones = number
    return hundreds, tens, ones

# How can you make this more scalable and reusable later?
def find_armstrong_numbers(numbers):
    ret = []
    
    # only need up to 3 digits of armstrong numbers, so only up to power of 3
    digits_power_one = [x for x in range(10)]
    digits_power_two = [x**2 for x in range(10)]
    digits_power_three = [x**3 for x in range(10)]
    
    # Goal is to take every digit of a number and add up
    # single digits - the digit
    # double digits - add the square of the digits
    # triple digits - add the cube of the digits
    for numb in numbers:
        hundreds, tens, ones = get_digits(numb)
        if hundreds != 0:
            calculated_sum = digits_power_three[hundreds] + digits_power_three[tens] + digits_power_three[ones]
        elif tens != 0:
            calculated_sum = digits_power_two[tens] + digits_power_two[ones]
        else:
            calculated_sum = digits_power_one[ones]
        if calculated_sum == numb:
            ret.append(numb)
    return ret